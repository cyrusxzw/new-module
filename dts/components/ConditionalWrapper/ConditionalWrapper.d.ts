import { FC } from 'react';
import type { ConditionalWrapperProps } from './ConditionalWrapper.types.';
declare const ConditionalWrapper: FC<ConditionalWrapperProps>;
export { ConditionalWrapper };
