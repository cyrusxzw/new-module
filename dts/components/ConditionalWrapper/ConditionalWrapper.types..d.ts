import { ReactNode, ReactElement } from 'react';
declare type Wrapper = (children: ReactNode) => ReactElement;
declare type ConditionalWrapperProps = {
  /**
   * If the condition is false, this functional component will be used instead of simply returning the children unwrapped.
   */
  alternateWrapper?: Wrapper;
  /**
   * If this boolean prop resolves to true the children prop will be wrapped in the wrapper prop.
   */
  condition: boolean;
  /**
   * A functional component that receives the children prop as an argument, `children => <div>{children}</div>`
   */
  wrapper: Wrapper;
};
export type { ConditionalWrapperProps };
