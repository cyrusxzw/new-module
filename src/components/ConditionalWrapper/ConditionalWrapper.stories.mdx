import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { ConditionalWrapper } from './ConditionalWrapper';

<Meta title="Components/ConditionalWrapper" component={ConditionalWrapper} />

# ConditionalWrapper

<Canvas>
  <Story name="Base component">
    <ConditionalWrapper
      condition={true}
      wrapper={(children) => <div>{children}</div>}
    >
      <span>Conditionally wrapped content</span>
    </ConditionalWrapper>
  </Story>
</Canvas>

## API

<ArgsTable of={ConditionalWrapper} />
