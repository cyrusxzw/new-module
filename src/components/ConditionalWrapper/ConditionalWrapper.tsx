import React from 'react';
import type { ConditionalWrapperType } from './ConditionalWrapper.types';

const ConditionalWrapper: ConditionalWrapperType = ({
  alternateWrapper,
  children,
  condition,
  wrapper,
}) => {
  if (condition) return wrapper(children);

  return alternateWrapper ? alternateWrapper(children) : <>{children}</>;
};

export { ConditionalWrapper };
