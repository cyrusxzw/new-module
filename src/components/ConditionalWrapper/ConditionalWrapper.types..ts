import { ReactNode, ReactElement } from 'react';

type Wrapper = (children: ReactNode) => ReactElement;

type ConditionalWrapperProps = {
  /**
   * If the condition is false, this functional component will be used instead of simply returning the children unwrapped.
   */
  alternateWrapper?: Wrapper;
  /**
   * If this boolean prop resolves to true the children prop will be wrapped in the wrapper prop.
   */
  condition: boolean;
  /**
   * A functional component that receives the children prop as an argument, `children => <div>{children}</div>`
   */
  wrapper: Wrapper;
};

export type { ConditionalWrapperProps };
