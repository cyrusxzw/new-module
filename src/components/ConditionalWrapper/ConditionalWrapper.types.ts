import type { ReactNode, ReactElement } from 'react';
import type { ComponentWithChildren } from '~/types';

type Wrapper = (children: ReactNode) => ReactElement;

type ConditionalWrapperProps = {
  /**
   * If the condition is false, this functional component will be used instead of simply returning the children unwrapped.
   */
  alternateWrapper?: Wrapper;
  /**
   * If this boolean prop resolves to true the children prop will be wrapped in the wrapper prop.
   */
  condition: boolean;
  /**
   * A functional component that receives the children prop as an argument, `children => <div>{children}</div>`
   */
  wrapper: Wrapper;
};

type ConditionalWrapperType = ComponentWithChildren<ConditionalWrapperProps>;

export type { ConditionalWrapperProps, ConditionalWrapperType };
