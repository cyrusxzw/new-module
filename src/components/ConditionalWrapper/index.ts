import { ConditionalWrapper } from './ConditionalWrapper';

export { ConditionalWrapper };
