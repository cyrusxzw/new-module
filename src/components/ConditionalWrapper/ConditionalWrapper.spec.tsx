import React from 'react';
import { render } from '@testing-library/react';
import { ConditionalWrapper } from './ConditionalWrapper';

describe('<ConditionalWrapper />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <ConditionalWrapper
        condition={true}
        wrapper={(children) => <div className="wrapper">{children}</div>}
      >
        <span className="children">children</span>
      </ConditionalWrapper>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders children only is condition is false', () => {
    const { container } = render(
      <ConditionalWrapper
        condition={false}
        wrapper={(children) => <div className="wrapper">{children}</div>}
      >
        <span className="children">children</span>
      </ConditionalWrapper>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders alternateWrapper correctly', () => {
    const { container } = render(
      <ConditionalWrapper
        alternateWrapper={(children) => (
          <div className="alternateWrapper">{children}</div>
        )}
        condition={false}
        wrapper={(children) => <div className="wrapper">{children}</div>}
      >
        <span className="children">children</span>
      </ConditionalWrapper>,
    );

    expect(container).toMatchSnapshot();
  });
});
