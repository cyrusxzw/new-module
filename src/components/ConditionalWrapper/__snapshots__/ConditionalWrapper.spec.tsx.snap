// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConditionalWrapper /> renders alternateWrapper correctly 1`] = `
<div>
  <div
    class="alternateWrapper"
  >
    <span
      class="children"
    >
      children
    </span>
  </div>
</div>
`;

exports[`<ConditionalWrapper /> should render base component correctly 1`] = `
<div>
  <div
    class="wrapper"
  >
    <span
      class="children"
    >
      children
    </span>
  </div>
</div>
`;

exports[`<ConditionalWrapper /> renders children only is condition is false 1`] = `
<div>
  <span
    class="children"
  >
    children
  </span>
</div>
`;
